File: java/src/com/paragonintel/codingexercise/Airports/AirportCollection.java
package com.paragonintel.codingexercise.Airports;

import com.google.gson.Gson;
import com.paragonintel.codingexercise.Location.GeoCoordinate;
import net.sf.javaml.core.kdtree.KDTree;

import java.io.*;

public class AirportCollection {

    KDTree airports;  // Use a KDTree to lookup nearest airport in log(n) time

    public AirportCollection(Airport[] airportArray) {
        airports = new KDTree(2);
        for (Airport airport : airportArray) {
            double[] coordinates = new double[2];
            coordinates[0] = airport.getLatitude();
            coordinates[1] = airport.getLongitude();
            airports.insert(coordinates, airport);
        }
    }

    public static AirportCollection loadFromFile(String filePath) throws IOException {
        var file = new File(filePath);
        if (!file.exists()) {
            throw new FileNotFoundException("File not found: " + filePath);
        }

        Reader reader = new FileReader(filePath);
        Gson gson = new Gson();
        Airport[] airports = gson.fromJson(reader, Airport[].class);
        return new AirportCollection(airports);
    }

    public Airport getClosestAirport(GeoCoordinate coordinate) {
        double[] latLongCoordinate = new double[2];
        latLongCoordinate[0] = coordinate.getLatitude();
        latLongCoordinate[1] = coordinate.getLongitude();
        return (Airport) airports.nearest(latLongCoordinate);
    }
}

File: csharp/Airports/AirportCollection.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Text.Json;

namespace ParagonCodingExercise.Airports
{
    public class AirportCollection
    {
        
        public AirportCollection(List<Airport> airports)
        {
            throw new NotImplementedException();
        }

        public static AirportCollection LoadFromFile(string filePath)
        {
            if (!File.Exists(filePath))
            {
                throw new FileNotFoundException("File not found", filePath);
            }

            using TextReader reader = new StreamReader(filePath);
            var json = reader.ReadToEnd();

            var airports = JsonSerializer.Deserialize<List<Airport>>(json);
            return new AirportCollection(airports);
        }

        public Airport GetClosestAirport(GeoCoordinate coordinate)
        {
            throw new NotImplementedException();
        }
    }
}
